android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/BaseSearchTelemetry.kt

```diff
@@ -25,17 +31,36 @@ import org.json.JSONObject
 /**
  * Main configuration and functionality for tracking ads / web searches with specific providers.
  */
-abstract class BaseSearchTelemetry {
+abstract class BaseSearchTelemetry(private val context: Context) {
+    private var providerList: List<SearchProviderModel>? = emptyList()
+    private var initializationDeferred = CompletableDeferred<Unit>()
+
+
+    private fun initializeProviderList() {
+        val parser = SERPTelemetryJsonParser(context)
+
+        CoroutineScope(Dispatchers.IO).launch {
+            providerList = parser.getSearchProviders()
+            initializationDeferred.complete(Unit)
+        }
+    }
+
+    internal fun getProviderForUrl(url: String): SearchProviderModel? {
+        if (!initializationDeferred.isCompleted || providerList.isNullOrEmpty()) {
+            initializeProviderList()
+
+            runBlocking {
+                initializationDeferred.await()
+            }
```

Comment:
Unfortunately I think we're definitely going to have to remove this `runBlocking` call. Since we're installing the subclasses of this during app startup, we can't risk blocking the main thread while we wait for the underlying network IO to complete. 

Because of that network IO, there's just no good way to make sure these installations happen synchronously anymore. My suggestion would be to make `abstract fun install` suspending, which would propogate through the children. Then, in `Components`, we can launch a coroutine that will complete the installation once this networking is finished. It should help remove the `Deferred` logic you've got here as well.